connect-inject/cleanup_resource_test.go

```diff
@@ -0,0 +1,322 @@
+			ConsulServices:      []capi.AgentServiceRegistration{consulFooSvc, consulFooSvcSidecar},
+			KubePods:            nil,
+			ExpConsulServiceIDs: nil,
+		},
+		"only app is registered, no sidecar": {
+			ConsulServices:      []capi.AgentServiceRegistration{consulFooSvc},
+			KubePods:            nil,
+			ExpConsulServiceIDs: nil,
+		},
+		"only sidecar is registered, no app": {
+			ConsulServices:      []capi.AgentServiceRegistration{consulFooSvcSidecar},
+			KubePods:            nil,
+			ExpConsulServiceIDs: nil,
+		},
+		"multiple instances of the same service": {
+			ConsulServices: []capi.AgentServiceRegistration{
+				consulFooSvc,
+				consulFooSvcSidecar,
+				consulFooSvcPod2,
+				consulFooSvcSidecarPod2,
+			},
+			KubePods:            []runtime.Object{fooPod},
+			ExpConsulServiceIDs: []string{"foo-abc123-foo", "foo-abc123-foo-sidecar-proxy"},
+		},
+	}
+
+	for name, c := range cases {
+		t.Run(name, func(t *testing.T) {
+			require := require.New(t)
+
+			// Start Consul server.
+			server, err := testutil.NewTestServerConfigT(t, nil)
+			defer server.Stop()
+			require.NoError(err)
+			server.WaitForLeader(t)
+			consulClient, err := capi.NewClient(&capi.Config{Address: server.HTTPAddr})
+			require.NoError(err)
+
+			// Register Consul services.
+			for _, svc := range c.ConsulServices {
+				require.NoError(consulClient.Agent().ServiceRegister(&svc))
+			}
+
+			// Create the cleanup resource.
+			log := hclog.Default().Named("cleanupResource")
+			log.SetLevel(hclog.Debug)
+			consulURL, err := url.Parse("http://" + server.HTTPAddr)
+			require.NoError(err)
+			kubeResources := c.KubePods
+			if !c.OutOfClusterNode {
+				node := nodeName(t, consulClient)
+				// NOTE: we need to add the node because the reconciler checks if
+				// the node the service is registered with actually exists in this
+				// cluster.
+				kubeResources = append(kubeResources, &corev1.Node{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: node,
+					},
+				})
+
+			}
+			cleanupResource := CleanupResource{
+				Log:              log,
+				KubernetesClient: fake.NewSimpleClientset(kubeResources...),
+				ConsulClient:     consulClient,
+				ConsulScheme:     consulURL.Scheme,
+				ConsulPort:       consulURL.Port(),
+			}
+
+			// Run Reconcile.
+			cleanupResource.reconcile()
+
+			// Test that the remaining services are what we expect.
+			services, err := consulClient.Agent().Services()
+			require.NoError(err)
+			var actualServiceIDs []string
+			for id := range services {
+				actualServiceIDs = append(actualServiceIDs, id)
+			}
+			require.ElementsMatch(actualServiceIDs, c.ExpConsulServiceIDs)
+		})
+	}
+}
+
+func TestDelete(t *testing.T) {
+	t.Parallel()
+
+	cases := map[string]struct {
+		Pod                 *corev1.Pod
+		ConsulServices      []capi.AgentServiceRegistration
+		ExpConsulServiceIDs []string
+		ExpErr              string
+	}{
+		"pod is nil": {
+			Pod:    nil,
+			ExpErr: "object for key default/foo was nil",
+		},
```

Comment:
Do we need a test that checks that we return an error if casting was not successful?

Reply:
Yep good catch.